### PR TITLE
Adding support for Polars structs

### PR DESCRIPTION
**Pull request by ayax79**
Provides support for reading Polars structs. This allows opening of supported files (jsonl, parquet, etc) that contain rows with structured data. 

The following attached json lines file([receipts.jsonl.gz](https://github.com/nushell/nushell/files/13311476/receipts.jsonl.gz)) contains a customer column with structured data. This json lines file can now be loaded via `dfr open` and will render as follows:

Screenshot 2023-11-09 at 10 09 18

This also addresses some cleanup of date handling and utilizing timezones where provided.

This pull request only addresses reading data from polars structs. I will address converting nushell data to polars structs in a future request as this change is large enough as it is.

### CODE REVIEW

**Comment by fdncred** [comment]
Oh, this looks interesting. I'm excited to see where it goes.

**Comment by fdncred** [comment]
Would you mind adding some examples of how this would be used, not in the code necessarily but just in the description?

Also, is this ready to land after review?

**Comment by ayax79** [comment]
> Would you mind adding some examples of how this would be used, not in the code necessarily but just in the description? Maybe this is just for opening things like parquet files that support all the polars data types?
> 
> Also, is this ready to land after review?

It's ready to go. 

The primary case with structs is one case. The way that I have been using it is with json lines files. For example, at my company I recently designed a system for tracing the life cycle of survey throughout our platform. It is pretty convenient to dump all json bodies into a json lines file and load that via `dfr open`. 

I have attached a very simple json lines file that renders the following results:
Screenshot 2023-11-09 at 10 09 18
[receipts.jsonl.gz](https://github.com/nushell/nushell/files/13311464/receipts.jsonl.gz)




> Would you mind adding some examples of how this would be used, not in the code necessarily but just in the description? Maybe this is just for opening things like parquet files that support all the polars data types?
> 
> Also, is this ready to land after review?

I updated the description. Yes this is ready for review.

**Comment by ayax79** [comment]
> Would you mind adding some examples of how this would be used, not in the code necessarily but just in the description? Maybe this is just for opening things like parquet files that support all the polars data types?
> 
> Also, is this ready to land after review?

Yes,  this is ready for review. I updated the description. Does that work?

**Review by fdncred — APPROVED** [review]
I tested and it works as expected. I'm good with these changes.

**Comment by fdncred** [comment]
Thanks!

File: crates/nu-cmd-dataframe/src/dataframe/values/nu_dataframe/conversion.rs [hunk 1]
```diff
@@ -996,10 +955,252 @@ fn series_to_values(
...
+            let arr = &*(arr as *const dyn Array as *const $casttype);
+            arr.value_unchecked(idx)
+        }};
+    }
+
+    // Not loving the unsafe here, however this largely based off the one
+    // example I found for converting Array values in:
+    // polars_core::chunked_array::ops::any_value::arr_to_any_value
+    unsafe {
+        match dt {
```
Comment (sholderbach) [inline crates/nu-cmd-dataframe/src/dataframe/values/nu_dataframe/conversion.rs:1061]: As the safety condition depends on the correct `DataType` being passed, this function itself has to be `unsafe`.

Even if we declare it as such I really find it hard to validate that this whole operation is sound. 
The call sites of this function need to guarantee both a valid type is ascribed and that all requirements with regards to the pointer and index are met.

Beyond that I am not super comfortable with a manul type reinterpretation that depends on the data representation as given to us by an external crate in this case both `polars_core` and `arrow2`. The former has been quite willing to break APIs so it feels to me like a ticking time bomb to make safety assumptions across crates outside of a simpler well documented interface.
If I have to read the crate source code to reassure me that this is sound and can't just read the documentation of a constrained `unsafe` API I think this is just asking for future trouble.

**Pull request merged**